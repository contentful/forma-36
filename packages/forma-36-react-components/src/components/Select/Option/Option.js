import React from 'react';
import PropTypes from 'prop-types';

export class Option extends React.Component {
  static propTypes = {
    value: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
  };

  render() {
    const { value, children, ...otherProps } = this.props;

    return (
      <option value={value} {...otherProps}>
        {children}
      </option>
    );
  }
}

export default Option;
