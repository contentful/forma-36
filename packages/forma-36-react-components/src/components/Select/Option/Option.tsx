import React from 'react';

export interface OptionProps {
  value: string;
  children: React.ReactNode;
  testId?: string;
}

const defaultProps: Partial<OptionProps> = {
  testId: 'cf-ui-select-option',
};

export const Option = (props: OptionProps) => {
  const { value, children, testId, ...otherProps } = props;

  return (
    <option value={value} data-test-id={testId} {...otherProps}>
      {children}
    </option>
  );
};

Option.defaultProps = defaultProps;

export default Option;
