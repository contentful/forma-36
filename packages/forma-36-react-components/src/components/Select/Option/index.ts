export * from './Option';
export { default } from './Option';
