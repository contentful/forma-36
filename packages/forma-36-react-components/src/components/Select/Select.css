@import 'settings/colors';
@import 'settings/dimensions';
@import 'settings/transitions';
@import 'settings/shadows';
@import 'settings/borders';

.Select__wrapper {
  position: relative;
  display: block;
}

.Select {
  width: 100%;
  display: block;
  appearance: none;
  padding: calc(1rem * (10 / var(--font-base-default)));
  padding-right: calc(1rem * (39 / var(--font-base-default)));
  background-color: var(--color-white);
  color: var(--color-text-mid);
  font-size: var(--font-size-m);
  font-family: var(--font-stack-primary);
  border: 1px solid var(--color-element-mid);
  border-radius: var(--border-radius-medium);
  box-shadow: var(--inset-box-shadow-default);
  height: calc(1rem * (40 / var(--font-base-default)));
  &:focus {
    outline: 0;
  }
  composes: focus-border--default from './../../styles/settings/a11y.css';
  cursor: pointer;

  &:-moz-focusring {
    color: transparent;
    text-shadow: 0 0 0 #000;
  }

  &::-ms-expand {
    display: none;
  }
}

.Select--disabled {
  background-color: var(--color-element-lightest);
  cursor: not-allowed;
  color: var(--color-text-light);

  &:focus,
  &:active {
    border-color: var(--color-element-mid);
  }
}

.Select--small {
  width: var(--input-width-small);
}

.Select--medium {
  width: var(--input-width-medium);
}

.Select--large {
  width: var(--input-width-large);
}

.Select--full {
  width: 100%;
}

.Select--auto {
  width: auto;
}

.Select--negative {
  border-color: var(--color-negative);
  &:focus {
    border-color: var(--color-negative);
    box-shadow: var(--glow-negative);
  }
  &.Select--disabled {
    border-color: var(--color-red-light);
  }
}

.Select__icon {
  position: absolute;
  right: calc(1rem * (8 / var(--font-base-default)));
  top: 50%;
  margin-top: -8px;
  pointer-events: none;
}
