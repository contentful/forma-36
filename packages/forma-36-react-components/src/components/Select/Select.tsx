import React, {
  useState,
  useEffect,
  ChangeEventHandler,
  FocusEventHandler,
  KeyboardEvent,
  ChangeEvent,
} from 'react';
import cn from 'classnames';
import Icon from '../Icon';
import styles from './Select.css';

export interface SelectProps {
  required?: boolean;
  name?: string;
  id?: string;
  hasError?: boolean;
  value?: string;
  isDisabled?: boolean;
  onChange?: ChangeEventHandler<HTMLSelectElement>;
  onBlur?: FocusEventHandler<HTMLSelectElement>;
  onFocus?: FocusEventHandler<HTMLSelectElement>;
  width?: 'auto' | 'small' | 'medium' | 'large' | 'full';
  testId?: string;
  className?: string;
  children: React.ReactNode;
  willBlurOnEsc?: boolean;
}

const defaultProps: Partial<SelectProps> = {
  testId: 'cf-ui-select',
  required: false,
  hasError: false,
  isDisabled: false,
  width: 'full',
  willBlurOnEsc: true,
};

export const Select = (props: SelectProps) => {
  const {
    id,
    name,
    required,
    children,
    width,
    className,
    testId,
    onChange,
    onBlur,
    onFocus,
    isDisabled,
    hasError,
    value,
    willBlurOnEsc,
    ...otherProps
  } = props;
  const [valueState, setValueState] = useState<string | undefined>();

  const handleKeyDown = (e: KeyboardEvent<HTMLSelectElement>) => {
    const ESC = 27;

    if (e.keyCode === ESC && willBlurOnEsc) {
      e.currentTarget.blur();
    }
  };

  const handleChange = (e: ChangeEvent<HTMLSelectElement>) => {
    if (!isDisabled) {
      setValueState(e.target.value);

      if (onChange) {
        onChange(e);
      }
    }
  };

  useEffect(() => {
    setValueState(value);
  }, [value]);

  const widthClass = `Select--${width}`;
  const classNames = cn(styles['Select'], {
    [styles['Select--disabled']]: isDisabled,
    [styles['Select--negative']]: hasError,
  });

  const wrapperClassNames = cn(
    styles['Select__wrapper'],
    styles[widthClass],
    className,
  );

  return (
    <div className={wrapperClassNames}>
      <select
        id={id}
        required={required}
        name={name}
        aria-label={name}
        data-test-id={testId}
        className={classNames}
        value={valueState}
        disabled={isDisabled}
        onFocus={onFocus}
        onChange={handleChange}
        onBlur={onBlur}
        onKeyDown={handleKeyDown}
        {...otherProps}
      >
        {children}
      </select>
      <Icon className={styles['Select__icon']} icon="ArrowDown" color="muted" />
    </div>
  );
};

Select.defaultProps = defaultProps;

export default Select;
