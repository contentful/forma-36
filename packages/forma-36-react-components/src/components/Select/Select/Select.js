import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import Icon from '../../Icon/Icon';
import styles from './Select.css';

export class Select extends React.Component {
  static propTypes = {
    required: PropTypes.bool,
    children: PropTypes.node.isRequired,
    name: PropTypes.string,
    id: PropTypes.string,
    hasError: PropTypes.bool,
    value: PropTypes.string,
    isDisabled: PropTypes.bool,
    onChange: PropTypes.func,
    onBlur: PropTypes.func,
    testId: PropTypes.string,
    onFocus: PropTypes.func,
    extraClassNames: PropTypes.string,
    width: PropTypes.oneOf(['auto', 'small', 'medium', 'large', 'full']),
  };

  static defaultProps = {
    name: undefined,
    id: undefined,
    extraClassNames: undefined,
    testId: 'cf-ui-select',
    required: false,
    hasError: false,
    isDisabled: false,
    width: 'full',
    value: undefined,
    onBlur: () => {},
    onFocus: () => {},
    onChange: () => {},
  };

  state = {
    value: this.props.value,
  };

  componentWillReceiveProps(nextProps) {
    if (this.props.value !== nextProps.value) {
      this.setState({
        value: nextProps.value,
      });
    }
  }

  render() {
    const {
      id,
      name,
      required,
      children,
      width,
      extraClassNames,
      testId,
      onChange,
      onBlur,
      onFocus,
      isDisabled,
      hasError,
      ...otherProps
    } = this.props;

    const widthClass = `Select--${width}`;
    const classNames = cn(styles.Select, {
      [styles['Select--disabled']]: isDisabled,
      [styles['Select--negative']]: hasError,
    });

    const wrapperClassNames = cn(
      styles.Select__wrapper,
      styles[widthClass],
      extraClassNames,
    );

    return (
      <div className={wrapperClassNames}>
        <select
          id={id}
          required={required}
          name={name}
          aria-label={name}
          // but select doesn't support for attribute :(
          htmlFor={name}
          data-test-id={testId}
          className={classNames}
          value={this.state.value}
          disabled={isDisabled}
          onFocus={onFocus}
          onChange={e => {
            if (!isDisabled) {
              this.setState({
                value: e.target.value,
              });
              onChange(e);
            }
          }}
          onBlur={onBlur}
          {...otherProps}
        >
          {children}
        </select>
        <Icon
          extraClassNames={styles.Select__icon}
          icon="ArrowDown"
          color="muted"
        />
      </div>
    );
  }
}

export default Select;
