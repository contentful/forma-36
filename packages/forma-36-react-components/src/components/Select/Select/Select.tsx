import React, { Component, ChangeEventHandler, FocusEventHandler } from 'react';
import cn from 'classnames';
import Icon from '../../Icon/Icon';
import styles from './Select.css';

export type SelectProps = {
  required?: boolean;
  children: React.ReactNode;
  name?: string;
  id?: string;
  hasError?: boolean;
  value?: string;
  isDisabled?: boolean;
  onChange?: ChangeEventHandler;
  onBlur?: FocusEventHandler;
  testId?: string;
  onFocus?: FocusEventHandler;
  className?: string;
  width?: 'auto' | 'small' | 'medium' | 'large' | 'full';
} & typeof defaultProps;

export interface SelectState {
  value?: string;
}

const defaultProps = {
  testId: 'cf-ui-select',
  required: false,
  hasError: false,
  isDisabled: false,
  width: 'full',
};

export class Select extends Component<SelectProps, SelectState> {
  static defaultProps = defaultProps;

  state = {
    value: this.props.value,
  };

  componentWillReceiveProps(nextProps: SelectProps) {
    if (this.props.value !== nextProps.value) {
      this.setState({
        value: nextProps.value,
      });
    }
  }

  render() {
    const {
      id,
      name,
      required,
      children,
      width,
      className,
      testId,
      onChange,
      onBlur,
      onFocus,
      isDisabled,
      hasError,
      ...otherProps
    } = this.props;

    const widthClass = `Select--${width}`;
    const classNames = cn(styles['Select'], {
      [styles['Select--disabled']]: isDisabled,
      [styles['Select--negative']]: hasError,
    });

    const wrapperClassNames = cn(
      styles['Select__wrapper'],
      styles[widthClass],
      className,
    );

    return (
      <div className={wrapperClassNames}>
        <select
          id={id}
          required={required}
          name={name}
          aria-label={name}
          data-test-id={testId}
          className={classNames}
          value={this.state.value}
          disabled={isDisabled}
          onFocus={onFocus}
          onChange={e => {
            if (!isDisabled) {
              this.setState({
                value: e.target.value,
              });
              if (onChange) {
                onChange(e);
              }
            }
          }}
          onBlur={onBlur}
          {...otherProps}
        >
          {children}
        </select>
        <Icon
          className={styles['Select__icon']}
          icon="ArrowDown"
          color="muted"
        />
      </div>
    );
  }
}

export default Select;
