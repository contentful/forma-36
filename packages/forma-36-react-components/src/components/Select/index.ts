export { Select } from './Select';
export type { SelectProps } from './Select';
export { Option } from './Option';
export type { OptionProps } from './Option';
