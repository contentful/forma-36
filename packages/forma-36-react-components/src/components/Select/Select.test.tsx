import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import axe from '../../utils/axeHelper';
import Select from './Select';
import Option from './Option';

it('renders the component', () => {
  const { container } = render(
    <Select name="optionSelect" id="optionSelect">
      <Option value="optionOne">Option One</Option>
    </Select>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with an extra class name', () => {
  const { container } = render(
    <Select name="optionSelect" id="optionSelect" className="extraClassName">
      <Option value="optionOne">Option One</Option>
    </Select>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component in disabled state', () => {
  const { container } = render(
    <Select isDisabled name="optionSelect" id="optionSelect">
      <Option value="optionOne">Option One</Option>
    </Select>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with error', () => {
  const { container } = render(
    <Select hasError name="optionSelect" id="optionSelect">
      <Option value="optionOne">Option One</Option>
    </Select>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with a value', () => {
  const { container } = render(
    <Select value="value1" name="optionSelect" id="optionSelect">
      <Option value="optionOne">Option One</Option>
    </Select>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component as required', () => {
  const { container } = render(
    <Select name="optionSelect" id="optionSelect" required>
      <Option value="optionOne">Option One</Option>
    </Select>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('should not dispatch onChange if disabled', () => {
  const mockOnChange = jest.fn();
  render(
    <Select
      name="optionSelect"
      id="optionSelect"
      onChange={mockOnChange}
      isDisabled
    >
      <Option value="optionOne">Option One</Option>
    </Select>,
  );

  userEvent.selectOptions(screen.getByTestId('cf-ui-select'), ['optionOne']);
  expect(mockOnChange).not.toHaveBeenCalled();
});

it('should dispatch onChange', () => {
  const mockOnChange = jest.fn();
  render(
    <Select name="optionSelect" id="optionSelect" onChange={mockOnChange}>
      <Option value="optionOne">Option One</Option>
    </Select>,
  );

  userEvent.selectOptions(screen.getByTestId('cf-ui-select'), ['optionOne']);
  expect(mockOnChange).toHaveBeenCalled();
});

it('has no a11y issues', async () => {
  const { container } = render(
    <Select id="optionSelect" name="optionSelect">
      <Option value="optionOne">Option 1</Option>
    </Select>,
  );
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});
