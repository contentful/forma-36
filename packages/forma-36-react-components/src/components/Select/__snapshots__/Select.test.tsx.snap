// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  class="Select__wrapper Select--full"
>
  <select
    aria-label="optionSelect"
    class="Select"
    data-test-id="cf-ui-select"
    id="optionSelect"
    name="optionSelect"
  >
    <option
      data-test-id="cf-ui-select-option"
      value="optionOne"
    >
      Option One
    </option>
  </select>
  <svg
    class="Icon Icon--small Icon--muted Select__icon"
    data-test-id="cf-ui-icon"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
</div>
`;

exports[`renders the component as required 1`] = `
<div
  class="Select__wrapper Select--full"
>
  <select
    aria-label="optionSelect"
    class="Select"
    data-test-id="cf-ui-select"
    id="optionSelect"
    name="optionSelect"
    required=""
  >
    <option
      data-test-id="cf-ui-select-option"
      value="optionOne"
    >
      Option One
    </option>
  </select>
  <svg
    class="Icon Icon--small Icon--muted Select__icon"
    data-test-id="cf-ui-icon"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
</div>
`;

exports[`renders the component in disabled state 1`] = `
<div
  class="Select__wrapper Select--full"
>
  <select
    aria-label="optionSelect"
    class="Select Select--disabled"
    data-test-id="cf-ui-select"
    disabled=""
    id="optionSelect"
    name="optionSelect"
  >
    <option
      data-test-id="cf-ui-select-option"
      value="optionOne"
    >
      Option One
    </option>
  </select>
  <svg
    class="Icon Icon--small Icon--muted Select__icon"
    data-test-id="cf-ui-icon"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
</div>
`;

exports[`renders the component with a value 1`] = `
<div
  class="Select__wrapper Select--full"
>
  <select
    aria-label="optionSelect"
    class="Select"
    data-test-id="cf-ui-select"
    id="optionSelect"
    name="optionSelect"
  >
    <option
      data-test-id="cf-ui-select-option"
      value="optionOne"
    >
      Option One
    </option>
  </select>
  <svg
    class="Icon Icon--small Icon--muted Select__icon"
    data-test-id="cf-ui-icon"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
</div>
`;

exports[`renders the component with an extra class name 1`] = `
<div
  class="Select__wrapper Select--full extraClassName"
>
  <select
    aria-label="optionSelect"
    class="Select"
    data-test-id="cf-ui-select"
    id="optionSelect"
    name="optionSelect"
  >
    <option
      data-test-id="cf-ui-select-option"
      value="optionOne"
    >
      Option One
    </option>
  </select>
  <svg
    class="Icon Icon--small Icon--muted Select__icon"
    data-test-id="cf-ui-icon"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
</div>
`;

exports[`renders the component with error 1`] = `
<div
  class="Select__wrapper Select--full"
>
  <select
    aria-label="optionSelect"
    class="Select Select--negative"
    data-test-id="cf-ui-select"
    id="optionSelect"
    name="optionSelect"
  >
    <option
      data-test-id="cf-ui-select-option"
      value="optionOne"
    >
      Option One
    </option>
  </select>
  <svg
    class="Icon Icon--small Icon--muted Select__icon"
    data-test-id="cf-ui-icon"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
</div>
`;
